### PR TITLE
QoL Dispenser Update

### PR DESCRIPTION
**Pull request by Protonull**
This change has been made [at the request of DriftL0rd](https://www.reddit.com/r/civclassics/comments/cq34as/minor_suggestion_regarding_xp_bottles/ewu238l/?context=6), he noticed that xp can only be granted via players throwing xp bottles after creating a machine to auto dispense. The response from the community was to install mods, I object to the line of thinking that gameplay should be replaced with mods, and so this change gives players more freedom. It also fixes a funny consequence of orb xp being disabled, see the individual commit messages for more information.

### CODE REVIEW

File: src/main/java/com/civclassic/oldenchanting/OldEnchanting.java [hunk 1]
```diff
@@ -302,13 +306,41 @@ private void createXPBottles(Player player, int totalExp) {
...
-		if(noExp) {
-			ProjectileSource source = event.getEntity().getShooter();
-			if(source instanceof Player) {
+		if (noExp) {
+			ThrownExpBottle bottle = event.getEntity();
+			ProjectileSource source = bottle.getShooter();
+			if (source instanceof Player) {
 				Player shooter = (Player) source;
 				shooter.giveExp(xpPerBottle);
+				bottle.teleport(shooter.getEyeLocation());
```
Comment (Maxopoly) [inline src/main/java/com/civclassic/oldenchanting/OldEnchanting.java]: Why is this needed

File: src/main/java/com/civclassic/oldenchanting/OldEnchanting.java [hunk 2]
```diff
@@ -302,13 +306,41 @@ private void createXPBottles(Player player, int totalExp) {
...
+			}
+			else if (source instanceof BlockProjectileSource) {
+				List<Player> nearby = bottle.getNearbyEntities(0.9, 0.9, 0.9)
```
Comment (Maxopoly) [inline src/main/java/com/civclassic/oldenchanting/OldEnchanting.java]: Could easily make that bigger to something like 5 blocks, ideally it'd even be a config parameter

File: src/main/java/com/civclassic/oldenchanting/OldEnchanting.java [hunk 3]
```diff
@@ -302,13 +306,41 @@ private void createXPBottles(Player player, int totalExp) {
...
+						.map((entity) -> (Player) entity)
+						.collect(Collectors.toList());
+				if (nearby.isEmpty()) {
+					event.setShowEffect(false);
+					event.setExperience(0);
+					return;
+				}
+				Player closest = null;
+				double distance = Double.MAX_VALUE;
+				for (Player player : nearby) {
```
Comment (Maxopoly) [inline src/main/java/com/civclassic/oldenchanting/OldEnchanting.java:419]: If you are already doing fancy stream stuff, you might as well go all the way and do
```
.map((entity) -> (Player) entity)
.min(p ->  p.getLocation().distanceSquared(bottle.getLocation()))
```

which gives you an `Optional<Player>`, circumventing the need for both the collector and your for loop.

**Review by Maxopoly — CHANGES_REQUESTED** [review]

File: src/main/java/com/civclassic/oldenchanting/OldEnchanting.java (hunk 3 above)
Comment (Maxopoly) [reply src/main/java/com/civclassic/oldenchanting/OldEnchanting.java:419]: In that case you could even do some obvious particle effect if the Optional is null to indicate that the XP is being wasted right now

File: src/main/java/com/civclassic/oldenchanting/OldEnchanting.java (hunk 1 above)
Comment (Protonull) [reply src/main/java/com/civclassic/oldenchanting/OldEnchanting.java]: When players throw Bottles o' Enchanting, they receive the xp on impact, regardless of the impact's location relative to the player. It means that the orbs themselves play no part in the player's leveling up, so the orbs will linger at the impact site until they expire or picked up.

This change teleports the orbs to the player at the moment of impact and in doing so removing them from the game but also providing the player with an audible feedback for receiving the xp.

File: src/main/java/com/civclassic/oldenchanting/OldEnchanting.java (hunk 2 above)
Comment (Protonull) [reply src/main/java/com/civclassic/oldenchanting/OldEnchanting.java]: Done, the radius is now a config option with the recommended settings being in between two and seven

**Comment by Protonull** [comment]
Just added the feature we discussed in private, you may now right click emeralds to level up
